asf_public_discourse_home_decarbonisation/analysis/mse/initial_text_analysis_category_data.py

```diff
@@ -0,0 +1,447 @@
+        "ngrams_col": "tokens_text",
+        "var_used": "posts and replies",
+        "min_frequency": min_frequency_tokens,
+    },
+    "bigrams_text": {
+        "data": lambda data: data,
+        "ngrams_col": "2_grams_text",
+        "var_used": "posts and replies",
+        "min_frequency": min_frequency_bigrams,
+    },
+    "trigrams_text": {
+        "data": lambda data: data,
+        "ngrams_col": "3_grams_text",
+        "var_used": "posts and replies",
+        "min_frequency": min_frequency_trigrams,
+    },
+    "tokens_text_op": {
+        "data": lambda data: data[data["is_original_post"] == 1],
+        "ngrams_col": "tokens_text",
+        "var_used": "posts",
+        "min_frequency": min_frequency_tokens,
+    },
+    "bigrams_text_op": {
+        "data": lambda data: data[data["is_original_post"] == 1],
+        "ngrams_col": "2_grams_text",
+        "var_used": "posts",
+        "min_frequency": min_frequency_bigrams,
+    },
+    "trigrams_text_op": {
+        "data": lambda data: data[data["is_original_post"] == 1],
+        "ngrams_col": "3_grams_text",
+        "var_used": "posts",
+        "min_frequency": min_frequency_trigrams,
+    },
+}
+
+
+def process_text_and_titles(mse_data: pd.DataFrame) -> pd.DataFrame:
+    """
+    Apply basic processing to text data and identifying stopwords.
+
+    Args:
+        mse_data (pd.DataFrame): Money Saving Expert data
+
+    Returns:
+        pd.DataFrame: MSE data with processed text
+    """
+    mse_data["processed_text"] = mse_data["text"].apply(lambda x: process_text(x))
+    mse_data["processed_title"] = mse_data["title"].apply(lambda x: process_text(x))
+    return mse_data
+
+
+def create_a_mapping_between_tokens_and_lemmas(mse_data: pd.DataFrame):
+    """
+    Create a mapping between tokens and lemmas for each token in either text or titles
+    Args:
+        mse_data (pd.DataFrame): _description_
+
+    Returns:
+        _type_: _description_
+    """
+    # Create a list of tokens in titles (removing repetitions from replies, which have the same title)
+    tokens_title = mse_data[mse_data["is_original_post"] == 1][["tokens_title"]]
+    tokens_title = tokens_title.explode("tokens_title").rename(
+        columns={"tokens_title": "tokens"}
+    )
+
+    # Create a list of tokens in text
+    tokens_text = mse_data[["tokens_text"]]
+    tokens_text = tokens_text.explode("tokens_text").rename(
+        columns={"tokens_text": "tokens"}
+    )
+
+    # Concatenate both lists
+    all_tokens = pd.concat([tokens_text, tokens_title])
+
+    # Unique tokens
+    all_tokens = all_tokens.drop_duplicates("tokens")
+
+    # Removing digits as they cannot be lemmatized
+    # (needs to explicitly to have `== False` because is_digit can also be missing)
+    all_tokens["is_digit"] = all_tokens["tokens"].str.isdigit()
+    all_tokens = all_tokens[all_tokens["is_digit"] == False]
+
+    lemmas_dictionary = lemmatize_sentence(list(all_tokens["tokens"]))
```

Comment:
Omg, I am so glad you caught this!

I was initially applying lemmatisation to each post/reply individually, but then because of the size of the dataset and the problems we we're experiencing I put everything together and completely forgot about this. But it's true, this will probably end up causing problems and it's obviously wrong.

Actually, even my original approach wasn't great. I was doing the lemmatisation to tokens in a post after removing tokens - but what I should be doing instead is lemmatising each sentence separately and then put all the tokens together. It's just tricky with the size of the dataset obviously.